Fixes and improvements in SOLPSSimulation, SOLPSMesh and related functions

vsnever opened the pull request:
This PR addresses the issues #8, #20, #31, #33, #34, #36 and #37.

Issue #8 is addressed in the following way:

- Safe to use `SOLPSFunction2D` and `SOLPSVectorFunction2D` classes are implemented. They replace  `SOLPSFunction3D` and `SOLPSVectorFunction3D`. For 3D, use `AxisymmetricMapper(SOLPSFunction2D)` and `VectorAxisymmetricMapper(SOLPSVectorFunction2D)`.

- `SOLPSMesh` and `SOLPSSimulation` can now be initialised without modification of hidden attributes;

- `SOLPSMesh` has new attributes for basis vectors, cell connection areas, indices of neighbouring cells and new functions `to_poloidal()` and `to_cartesian()` for converting vectors defined on a grid from/to (poloidal, radial)/(R, Z);

- `species_list` in `SOLPSSimulation` is a list of (name, charge) pairs now instead of a list of strings;

- Setting `b_field` and `velocities` vectors in curvilinear poloidal coordinates (**e**<sub>pol</sub>, **e**<sub>rad</sub>, **e**<sub>tor</sub>) automaticaly sets `b_field_cylindrical` and `velocities_cylindrical` (**e**<sub>R</sub>, **e**<sub>φ</sub>, **e**<sub>Z</sub>) and vice versa (note: **e**<sub>tor</sub> = -**e**<sub>φ</sub>);

- Incorrect calculation of basis vectors using cell centres is fixed (basis vectors are defined by left and bottom faces of the cell);

- The code for `mesh.triangles` and `mesh.triangle_to_grid_map` creation is vectorised (as all other code wherever possible);

- `SOLPSSimulation` now has setters for its properties;

- `load_solps_from_mdsplus()`, `load_solps_from_raw_output()` and `load_solps_from_balance()` do not brake `SOLPSSimulation` interface anymore;
   
- `load_solps_from_mdsplus()` and `load_solps_from_raw_output()` are made more robust. Now both support B2 standalone runs. Error handling is added to `load_solps_from_mdsplus()` for some optional attributes that may not be present on the server;

- Other small improvements.

Issue #20 is fixed. `*_f2d()` and `*_f3d()` interpolators are added for plasma parameters. For vectors `_carteisan` suffix is used instead of `_f3d`.

Issue #31 is fixed. Now if the neutral densities from Eirene are available, they replace the neutral densities from B2.

Issue #33 is fixed using `lookup_isotope()` and `lookup_element()` methods. The plasma composition is defined from nuclear charge, atomic mass number and electrical charge. This is more robust than parsing a string with species names.

Issue #34 is addressed by parsing the following additional quantities in `load_solps_from_mdsplus()` and `load_solps_from_raw_output()`:

- Ion temperature;
- magnetic field (previously was read only from MDS);
- atom/ion fluxes (B2), neutral atom densities (Eirene, if available),
- neutral atom fluxes (Eirene, if available);
- neutral atom temperatures (Eirene, if available).

_The fluxes are not stored in `SOLPSSimulation`, but used to calculate velocities._

Issue #36 is fixed. The parallel velocity and the B2 fluxes (in s<sup>-1</sup>) defined on cell faces are used in `b2_flux_to_velocity()` to calculate velocities at cell centres. If neutral atom fluxes from Eirene (in m<sup>-2</sup> s<sup>-1</sup>) defined at cell centres are available, they are used in `eirene_flux_to_velocity()` replacing the B2 ones.

Issue #37 is fixed. Now all arrays are row-major. However, the indexing is inverted. The correct index order now is: (radial, poloidal) or (species, radial, poloidal) for scalars and (vector component, radial, poloidal) or (species, vector component, radial, poloidal) for vectors. This should improve the performance in typical workflows.

**Remaining issues:**

- Electron velocities are still not read from SOLPS data.

- Molecular and total H-alpha emission are still not read from SOLPS data.

vsnever committed:
Removed _popular_species and _species_symbol_map dicts, added setters to SOLPSSimulation, other fixes.

vsnever committed:
Added ion and nuetral atom temperatures.

vsnever commented:
Just updated the code by adding ion and neutral atom temperatures. The latter is not available in stand-alone B2.5 simulations.

vsnever committed:
Many fixes and improvements in SOLPSMesh class

vsnever commented:
I renamed this PR because now it addresses not just #31 but in some degree #8 too.

Replacing the algorithm for poloidal basis vectors calculation in `SOLPSMesh` was more a bug fix than an improvement. The SOLPS mesh have break points due to the separatrix. In the old algorithm, which calculated the basis vectors using the cell centres, the basis vectors had wrong values at these break points.

vsnever committed:
Corrected mesh basis vector calculation

vsnever commented:
Just noticed that conversion from poloidal to Cartesian coordinates in `mdsplus.py` is incorrect. It's for orthogonal basis, but SOLPS uses curvilinear coordinates.

vsnever commented:
> Just noticed that conversion from poloidal to Cartesian coordinates in `mdsplus.py` is incorrect. It's for orthogonal basis, but SOLPS uses curvilinear coordinates.

Sorry, I'm wrong here. It converts **to** Cartesian, so it's OK.

vsnever committed:
Added poloidal <--> Cartesian conversion to SOLPSMesh. Added conversion on assignment to b_field and velocities in SOLPSSimulation.

vsnever committed:
Replaced nested loops in _map_data_onto_triangles with vector code.

mattngc commented:
Hey Vlad, I don't have any strong opinions on these improvements. I don't use this module as much at the moment, so I think opinions from the main users are more important. I'll let @jacklovell and @Mateasek give feedback on these changes.

vsnever committed:
Added magnetic files, velocity and neutral temperature reading to raw_simulation_files. Removed radial_particle_flux and radial_area attributes from SOLPSSimulation.

vsnever committed:
Added exception handling to load_solps_from_mdsplus() for better support of B2.5 stand-alone simulations

vsnever committed:
Renamed parallel direction to poloidal in SOLPSMesh to match SOLPS notations.

vsnever commented:
> Hey Vlad, I don't have any strong opinions on these improvements. I don't use this module as much at the moment, so I think opinions from the main users are more important. I'll let @jacklovell and @Mateasek give feedback on these changes.

Hi Matthew, thanks for reply. I need a couple more days to get this PR ready for review.

vsnever committed:
Added fluxes to velocity conversion to adress #36.

vsnever committed:
Removed modification of assigned SOLPSSimulation attributes to simplify future implementation of #20 via setters.

vsnever committed:
Added set_..() and get_..() interfaces to SOLPSSimulation to make easier future implementation of #20.

vsnever committed:
Made assign_fort44_parser() less strict to adress #32

vsnever committed:
Fixed errors with pickle save/load.

vsnever commented:
In the first post, I summaraised the changes made in this PR and marked it ready for review.

Hi @jacklovell and @Mateasek, I am awaiting your verdict).

Hello @jrh-ccfe, could you please share any balance.nc file, so I could improve `load_solps_from_balance()`. I have access to Heimdall cluster, so a path to the file will be enough.

CnlPepper commented on cherab/solps/eirene/parser/fort44.py:
```diff
@@ -57,13 +57,17 @@ def assign_fort44_parser(file_version):
     :param file_version: Fort44 file version from the file header.
     :return: Parsing function object
     """
+    fort44_supported_versios = (
```
typo in name: "versios" -> "versions"

CnlPepper commented on cherab/solps/eirene/parser/fort44.py:
```diff
@@ -57,13 +57,17 @@ def assign_fort44_parser(file_version):
-        20170328: load_fort44_2017,
-        20130210: load_fort44_2013
-    }
+    if file_version not in fort44_supported_versios:
```
versios -> versions

vsnever replied:
Thanks, this is corrected.

CnlPepper commented on cherab/solps/formats/balance.py:
```diff
@@ -93,65 +61,111 @@ def load_solps_from_balance(balance_filename):
-    inside_outside = AxisymmetricMapper(Discrete2DMesh(mesh.vertex_coords, mesh.triangles, inside_outside_data, limit=False))
-    sim._inside_mesh = inside_outside
+    # Load electron species
+    sim.set_electron_temperature(fhandle.variables['te'].data.copy() / el_charge)
```
shouldn't this be a property not a setter?

```
sim.electron_temperature = ....
```

CnlPepper replied:
ignore me it this needs to be optimised for speed.... but I suspect it doesn't need to be.

vsnever replied:
There is a discussion in #20 and also in comments to the `SOLPSSimulation` code, that `sim.electron_temperature`, etc. should return not the data arrays but the interpolators. However, this is not decided yet. If `sim.electron_temperature` return the interpolator, than setting the electron temperature data in the following style:
```
sim.electron_temperature = data
```
will lead to magic hat paradox, because the user puts one object there, but then gets the different object.

vsnever replied:
Probably the best thing would be to have different properties for data and interpolators, like
```
sim.electron_temperature_data  # for grid data
sim.electron_temperature  # for interpolator
```
@jacklovell, @Mateasek, what do you think?

CnlPepper replied:
I would implement it as you suggest in the most recent comment. If users wish to obtain either an interpolator or the raw data, they should be accessible via separate attributes.

I am curious why these values are not being set via the ```__init__``` rather than updating each attribute. Is there a plan for people to write data by creating an empty object first? If not it may be cleaner to configure the object via an init - like the Equilibrium class. i.e. raw data passed to init.

CnlPepper replied:
Actually ignore the last point, it is just style and we use a similar pattern for the Plasma object.

CnlPepper replied:
Ok I've just had a better look through the code to remind myself. I think there is some confusion here. The simulation object was just meant to be a thin wrapper holding the raw data. So all the properties should be data. The data gets converted to functions when you ask for a plasma object. @Mateasek was asking for additional methods / properties that return the data in functional form i.e. splitting out the code from the create_plasma() method. Can I suggest the following:

```
sim.electron_temperature             # property, raw data arrays
sim.electron_temperature_as_f2d()    # method, function 2d
sim.electron_temperature_as_f3d()    # method, function 3d
```

vsnever replied:
I like this interface. I will replace the `set _... ()` methods with property setters and remove the `get _... ()` methods. I suggest a separate PR for adding the interpolators, because we have to decide whether we use `Discrete2DMesh` and `AxisymmetricMapper` or `SOLPSFunction3D` (see my comment in #20).

CnlPepper replied:
I would use the ```SOLPSFunction3D``` for these. 

Btw I've always been very uncomfortable with the use of the ```AxisymmetricMapper``` and ```Discrete2DMesh``` functions in ```SOLPSFunction*```. Using a float to hold an integer is dangerous, it will (with absolute certainty) result in the wrong data being indexed. I would rewrite the ```SOLPSFunction*``` classes. Have a quick look at the ```Discrete2DMesh``` code:
https://github.com/raysect/source/blob/development/raysect/core/math/function/float/function2d/interpolate/discrete2dmesh.pyx

It is trivial to get it to just return the index. See line 174. I would use the KDtree directly in the SOLPS function, copying the relevant code from``` Discrete2DMesh```:

https://github.com/raysect/source/blob/development/raysect/core/math/function/float/function2d/interpolate/common.pyx

Make the SOLPS functions 2D and them use the axisymmetric wrapper around the SOLPSFunction2D to convert it to 3D, where the result is actually a double!

vsnever replied:
I implemented `SOLPSFunction2D` and `SOLPSVectorFunction2D` classes. Also added `*_f2d()` and `*_f3d()` interpolators for plasma parameters. The demos work, but more testing will probably be needed, which I'll leave for tomorrow.

CnlPepper replied:
Nice work. I see you went with properties due to the low overhead of instancing the existing inside function. All looks good, much cleaner and more reliable.

CnlPepper commented on cherab/solps/solps_plasma.py:
```diff
@@ -32,139 +31,219 @@
+        self._ion_temperature = value
 
     @property
-    def radial_particle_flux(self):
```
Setter and getters are not python style. These really should be properties:

```
@property
def myvar(self):
    ......

@myvar.setter
def myvar(self, value):
    ......
```

vsnever committed:
Corrected the typo

vsnever committed:
Made all array row-major and inverted indexing. Removed get_*(), set_*() methods, added setters instead.

vsnever committed:
Implemented SOLPSFunction2D and SOLPSVectorFunction2D, added *_f2d() and *_f3d() interpolators.

vsnever committed:
Updated __getstate__, __setstate__ in SOLPSSimulation.

vsnever committed:
Properly distinguished poloidal (poloidal, radial, toroidal), cylindrical (R, phi, Z) and Cartesian (x, y, z) coordinates, e_tor = -e_phi.

vsnever committed:
Added docstrings to new SOLPSSimulation attributes and some safety checks

vsnever committed:
Fixed a typo

vsnever committed:
Moved b2_flux_to_velocity() and eirene_flux_to_velocity() to SOLPSSimulation, fixed a bug in flux density calculation. Added electron_velocities_* attributes.

vsnever committed:
eirene.elosm and eirene.edism now resolved over molecules like in fort.44. Made fort.44 version check strict again.

vsnever committed:
Updated balance.py to match the functionality of mdsplus.py.

vsnever commented:
I updated balance.py and revert the change in the fort44 parser.

jacklovell commented on cherab/solps/b2/parse_b2_block_file.py:
```diff
@@ -47,16 +47,16 @@ def _make_solps_data_object(_data):
         # Multiple 2D data field (e.g. na)
         if number > nxyg:
-            _data = np.array(_data).reshape((nxg, nyg, int(number / nxyg)), order='F')
+            _data = np.array(_data).reshape((int(number / nxyg), nyg, nxg))
```
I would use floor division here, instead of manually doing float division and truncating to int. It'll produce cleaner code: compare `int(number / nxyg)` with `number // nxyg`.

jacklovell commented on cherab/solps/b2/parse_b2_block_file.py:
```diff
@@ -47,16 +47,16 @@ def _make_solps_data_object(_data):
+            _data = np.array(_data).reshape((int(number / nxyg), nyg, nxg))
             if debug:
-                print('Mesh data field {} with dimensions:  {:d} x {:d} x {:d}'.format(name, nxg, nyg, int(number/nxyg)))
+                print('Mesh data field {} with dimensions:  {:d} x {:d} x {:d}'.format(name, int(number / nxyg), nyg, nxg))
```
Same comment re: floor division.

jacklovell commented on cherab/solps/eirene/parser/fort44_2013.py:
```diff
@@ -16,6 +16,7 @@
 # See the Licence for the specific language governing permissions and limitations
 # under the Licence.
 
+import numpy as np
```
This numpy import is not needed.

jacklovell commented on cherab/solps/eirene/parser/fort44_2017.py:
```diff
@@ -16,6 +16,7 @@
 # See the Licence for the specific language governing permissions and limitations
 # under the Licence.
 
+import numpy as np
```
Also not needed.

jacklovell commented on cherab/solps/eirene/eirene.py:
```diff
@@ -91,7 +91,7 @@ def __init__(self, nx, ny, na, nm, ni, ns, species_labels, version=None,
     @property
```
There's also an unused numpy import in this file.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -19,139 +19,185 @@
+from scipy import constants
 from raysect.core.math.function.float import Discrete2DMesh
 
 from cherab.core.math.mappers import AxisymmetricMapper
```
These two imports are no longer needed.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -19,139 +19,185 @@
+        ns = am.size
+        for i in range(ns):
+            symbol = ''.join([b.decode('utf-8').strip(' 0123456789+-') for b in species_names[i]])  # also strips isotope number
+            if symbol != 'D' and symbol != 'T':
```
More Pythonic and extensible: `if symbol not in ('D', 'T')`

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -19,139 +19,185 @@
+        for i in range(ns):
+            symbol = ''.join([b.decode('utf-8').strip(' 0123456789+-') for b in species_names[i]])  # also strips isotope number
+            if symbol != 'D' and symbol != 'T':
+                isotope = lookup_isotope(symbol, number=am[i])  # will through an error for D or T
```
Typo: throw

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -19,139 +19,185 @@
+            # Save total radiated power to the simulation object
+            total_rad = rydberg_energy * el_charge * potential_loss - b2_ploss
+
+    if isinstance(total_rad, np.ndarray):
```
Setting sim.total_radiation will convert to an ndarray if it isn't already. I'm not sure what else this is checking. And the indentation means the line will also be called when `b2_standalone` is `True`, which will result in an `UnboundLocalError` as `total_rad` isn't even assigned in that case. So the indentation needs to be increased by 8 (I think).

vsnever replied:
`total_rad` is calculated even if `b2_standalone` is `True` (without neutrals), but the indentation was wrong indeed, this must be inside `with` statement. Corrected.

jacklovell commented on cherab/solps/formats/mdsplus.py:
```diff
@@ -17,20 +17,16 @@
-import re
 import numpy as np
-from math import sqrt
 from raysect.core import Point2D
```
Point2D is no longer used.

jacklovell commented on cherab/solps/formats/mdsplus.py:
```diff
@@ -40,215 +36,173 @@ def load_solps_from_mdsplus(mds_server, ref_number):
     """
 
-    from MDSplus import Connection as MDSConnection
+    from MDSplus import Connection as MDSConnection, mdsExceptions
```
`mdsExceptions` is only available in MDSplus 7, but at Culham we use MDSplus 6.1. You could use `MDSplus.MdsException` instead: it's compatible with both 6.1 and 7 and is a superclass of `TreeNNF`.

jacklovell commented on cherab/solps/formats/mdsplus.py:
```diff
@@ -40,215 +36,173 @@ def load_solps_from_mdsplus(mds_server, ref_number):
+    mesh = load_mesh_from_mdsplus(conn, mdsExceptions)
+
+    # Load each plasma species in simulation
+    ns = conn.get('\SOLPS::TOP.IDENT.NS').data()  # Number of species
```
While we're doing a tidy-up, these should be r-strings instead of standard strings. Otherwise Python tries to treat the backslash as an escape character. For robustness it should be:
`ns = conn.get(r'\SOLPS::TOP.IDENT.NS').data()`.

This is true for all the `conn.get` calls.

jacklovell commented on cherab/solps/formats/mdsplus.py:
```diff
@@ -40,215 +36,173 @@ def load_solps_from_mdsplus(mds_server, ref_number):
-    # Master list of species, e.g. ['D0', 'D+1', 'C0', 'C+1', ...
-    species_list = conn.get('\SOLPS::TOP.IDENT.SPECIES').data()
+    sim.b_field = conn.get('\SOLPS::TOP.SNAPSHOT.B').data()[:3]
+    # sim.b_field_cylindrical is created authomatically
```
Typo: automatically.

jacklovell commented on cherab/solps/formats/mdsplus.py:
```diff
@@ -40,215 +36,173 @@ def load_solps_from_mdsplus(mds_server, ref_number):
+
+    # Obtaining additional data from EIRENE and replacing data for neutrals
+
+    b2_standalone = False
```
I'd prefer this to become a `try/except/else` block, with `b2_standalone = False` set in the `else` section. It reads a bit more naturally this way: "try: reading Eirene data. exception: b2 standalone. No exception: not b2 stadalone".

jacklovell commented on cherab/solps/formats/mdsplus.py:
```diff
@@ -17,20 +17,16 @@
-from cherab.solps.solps_plasma import SOLPSSimulation
+from cherab.solps.solps_plasma import SOLPSSimulation, prefer_element
 
+from matplotlib import pyplot as plt
```
pyplot is unused.

jacklovell commented on cherab/solps/formats/raw_simulation_files.py:
```diff
@@ -19,33 +19,16 @@
+from scipy.constants import elementary_charge
 from raysect.core.math.function.float import Discrete2DMesh
 
 from cherab.core.math.mappers import AxisymmetricMapper
```
These imports are no longer used.

jacklovell commented on cherab/solps/formats/raw_simulation_files.py:
```diff
@@ -56,93 +39,154 @@ def load_solps_from_raw_output(simulation_path, debug=False):
 
     if not os.path.isdir(simulation_path):
-        RuntimeError("simulation_path must be a valid directory")
+        RuntimeError("Simulation_path must be a valid directory.")
```
`simulation_path` is the (case sensitive) name of the argument to this function. It should be left as is. Agree with the additional full stop though.

jacklovell replied:
Also, there's a missing `raise` statement here.

jacklovell commented on cherab/solps/formats/raw_simulation_files.py:
```diff
@@ -56,93 +39,154 @@ def load_solps_from_raw_output(simulation_path, debug=False):
 
     if not(os.path.isfile(b2_state_file)):
-        RuntimeError("No B2 b2fstate file found in SOLPS output directory")
+        RuntimeError("No B2 b2fstate file found in SOLPS output directory.")
```
Missing `raise`.

jacklovell commented on cherab/solps/formats/raw_simulation_files.py:
```diff
@@ -56,93 +39,154 @@ def load_solps_from_raw_output(simulation_path, debug=False):
-        raise RuntimeError("No B2 b2fgmtry file found in SOLPS output directory")
+        raise RuntimeError("No B2 b2fgmtry file found in SOLPS output directory.")
 
     if not(os.path.isfile(b2_state_file)):
```
Superfluous parentheses.

jacklovell commented on cherab/solps/formats/raw_simulation_files.py:
```diff
@@ -56,93 +39,154 @@ def load_solps_from_raw_output(simulation_path, debug=False):
-        RuntimeError("No B2 b2fstate file found in SOLPS output directory")
+        RuntimeError("No B2 b2fstate file found in SOLPS output directory.")
 
     if not(os.path.isfile(eirene_fort44_file)):
```
Superfluous parentheses.

jacklovell commented on cherab/solps/mesh_geometry.py:
```diff
@@ -18,16 +18,12 @@
 
 # External imports
-from collections import namedtuple
+# from collections import namedtuple
```
Delete rather than comment out.

jacklovell commented on cherab/solps/mesh_geometry.py:
```diff
@@ -40,89 +36,122 @@ class SOLPSMesh:
+    def __init__(self, r, z, vol, neighbix, neighbiy):
+
+        if r.shape != z.shape:
+            raise ValueError('Shape of r array: %s mismatch the shape of z array: %s.' % (r.shape, z.shape))
```
Please use string.format rather than %-style formatting, for consistency with the rest of the Cherab code base.

jacklovell commented on cherab/solps/mesh_geometry.py:
```diff
@@ -40,89 +36,122 @@ class SOLPSMesh:
+        self._poloidal_area = np.pi * (r[2] + r[0]) * vec_magn
+
+        # For convertion from Cartesian to poloidal
+        # TODO Make it work with trianle cells
```
Typo: triangle.

jacklovell commented on cherab/solps/mesh_geometry.py:
```diff
@@ -40,89 +36,122 @@ class SOLPSMesh:
+        # plt.quiver(self._cr[-1], self._cz[-1], self._radial_basis_vector[0, -1], self._radial_basis_vector[1, -1], color='k')
+        # plt.quiver(self._cr[-1], self._cz[-1], self._poloidal_basis_vector[0, -1], self._poloidal_basis_vector[1, -1], color='r')
+        # plt.gca().set_aspect('equal')
+        # plt.show()
```
Delete unused test code.

jacklovell commented on cherab/solps/mesh_geometry.py:
```diff
@@ -40,89 +36,122 @@ class SOLPSMesh:
     """
 
-    def __init__(self, cr_r, cr_z, vol):
+    # TODO Make neighbix and neighbix optional in the future, as they can be reconstructed with _tri_index_loopup
```
`_tri_index_lookup` no longer exists. Perhaps modify this comment to `_triangle_to_grid_map`.

jacklovell commented on cherab/solps/mesh_geometry.py:
```diff
@@ -135,70 +164,161 @@ def cz(self):
+        #     for i in range(self.vessel.shape[0]):
+        #         ax.plot([self.vessel[i, 0], self.vessel[i, 2]], [self.vessel[i, 1], self.vessel[i, 3]], 'k')
+        #     for i in range(self.vessel.shape[0]):
+        #         ax.plot([self.vessel[i, 0], self.vessel[i, 2]], [self.vessel[i, 1], self.vessel[i, 3]], 'or')
```
Delete commented out code, or un-comment it.

jacklovell commented on cherab/solps/solps_2d_functions.pyx:
```diff
@@ -0,0 +1,214 @@
+        if self._kdtree.is_contained(new_point2d(x, y)):
+
+            triangle_id = self._kdtree.triangle_id
+            ix, iy = self._triangle_to_grid_map_mv[triangle_id, :]
```
Tuple unpacking is unfortunately not optimised by Cython: this still goes through a high level Python object (see the HTML output of `cython -a solps_2d_functions.pyx`. A more efficient solution would be:
```python
indices = self._triangle_to_grid_map_mv[triangle_id, :]
ix = indices[0]
iy = indices[1]
```
Of course, `indices` would need to be `cdef`'d appropriately.

jacklovell commented on cherab/solps/solps_2d_functions.pyx:
```diff
@@ -0,0 +1,214 @@
+        if self._kdtree.is_contained(new_point2d(x, y)):
+
+            triangle_id = self._kdtree.triangle_id
+            ix, iy = self._triangle_to_grid_map_mv[triangle_id, :]
```
See previous comment on tuple unpacking.

jacklovell commented on cherab/solps/solps_3d_functions.pyx:
```diff
@@ -101,7 +101,9 @@ cdef class SOLPSVectorFunction3D(VectorFunction3D):
-        vx, vy, vz = self._grid_vectors[ix, iy, :]
+        vx = self._grid_vectors[0, ix, iy]
+        vy = self._grid_vectors[1, ix, iy]
+        vz = self._grid_vectors[2, ix, iy]
```
Do the same unrolling on the `ix, iy` assignments in both `evaluate` methods.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -17,228 +17,577 @@
-import re
 import pickle
 import numpy as np
 import matplotlib.pyplot as plt
```
pyplot is unused.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -17,228 +17,577 @@
 import numpy as np
 import matplotlib.pyplot as plt
-from scipy.constants import atomic_mass, electron_mass
+from scipy.constants import atomic_mass, electron_mass, elementary_charge
```
Elementary charge is unused: left over from development?

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -17,228 +17,577 @@
-from raysect.core.math.function.float import Discrete2DMesh
 from raysect.core import translate, Point3D, Vector3D, Node, AffineMatrix3D
 from raysect.primitive import Cylinder
 from raysect.optical import Spectrum
```
Of these, only `translate`, `Vector3D` and `Cylinder` are actually used. The rest can be removed.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -17,228 +17,577 @@
+        self._b_field_cartesian = None
+        self._eirene_model = None  # what is this for?
+        self._b2_model = None  # what is this for?
+        self._eirene = None  # do we need this in SOLPSSimulation?
```
I think it's fine to leave these here: they provide slightly more direct access to the underlying simulation data and can be used for debugging purposes.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -17,228 +17,577 @@
 
         This is not calculated from the CHERAB emission models, instead it comes from the SOLPS output data.
         Is calculated from the sum of all integrated line emission and all Bremmstrahlung. The signals used are 'RQRAD'
         and 'RQBRM'. Final output is in W/str?
```
I think this should be in W/m3, based on #28.

vsnever replied:
I corrected this to W m-3, however this docstring is valid only if `total_radiation` is read from MDSPlus server. When obtained from raw output, `total_radiation` includes only radiation from neutral atoms, when obtained from balance.nc, it does not include bremsstrahlung.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -17,228 +17,577 @@
-            raise RuntimeError("Total radiation not available for this simulation")
+        if self._total_radiation is None:
+            raise RuntimeError("Total radiation not available for this simulation.")
         else:
```
`else` is not necessary after a `raise`. This applies to all the `total_radiation*`, `b_field*` and `eirene_simulation` properties.

vsnever replied:
This code was before me, some properties threw errors when not initialized and some did not. I think, they should not throw errors. If a property is required for a function to work, that function must check if the value is set and throw a `RuntimeError` if not. I've removed these `raise` statements from properties and instead added them to functions that don't work without these properties.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -248,55 +597,201 @@ def eirene_simulation(self):
+        topiy = mesh.neighbiy[3]
+
+        # Converting s-1 --> m-2 s-1
+        poloidal_flux = np.divide(poloidal_flux, poloidal_area, out=np.zeros_like(poloidal_flux), where=poloidal_area > 0)
+        radial_flux = np.divide(radial_flux, radial_area, out=np.zeros_like(radial_flux), where=radial_area > 0)
```
Does specifying the `out` argument here actually help? You're still having to allocate a new array, and I thought the `out` argument was only relevant if there was an existing array to put the result in to avoid a new allocation. There are a few other places below where you also do this.

vsnever replied:
No, if I do not specify the `out` as a zero array explicitly, I'll get uninititalised values where the condition for divide is not met. From numpy.divide description: "...Note that if an uninitialized out array is created via the default out=None, locations within it where the condition is False will remain uninitialized.", [https://numpy.org/doc/stable/reference/generated/numpy.divide.html](https://numpy.org/doc/stable/reference/generated/numpy.divide.html).

jacklovell replied:
Ah, I see. Thanks for clarifying.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -460,77 +955,66 @@ def create_plasma(self, parent=None, transform=None, name=None):
-        for k, sp in enumerate(self.species_list):
+        if self.electron_velocities_cartesian is None:
+            print('Warning! No electron velocity data available for this simulation.')
+            electron_velocity = lambda x, y, z: Vector3D(0, 0, 0)
```
This should be a `ConstantVector3D` object, not a lambda function.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -460,77 +955,66 @@ def create_plasma(self, parent=None, transform=None, name=None):
+            if self.velocities_cartesian is not None:
+                velocity = self.velocities_cartesian[k]
             else:
                 velocity = lambda x, y, z: Vector3D(0, 0, 0)
```
This should be a `ConstantVector3D` object, not a lambda function.

jacklovell commented on cherab/solps/solps_plasma.py:
```diff
@@ -460,77 +955,66 @@ def create_plasma(self, parent=None, transform=None, name=None):
+        else:
+            electron_velocity = self.electron_velocities_cartesian
+        plasma.electron_distribution = Maxwellian(self.electron_density_f3d, self.electron_temperature_f3d, electron_velocity, electron_mass)
+ 
```
Trailing whitespace.

jacklovell reviewed (changes requested):
Thanks for this Vlad: it's a huge amount of great work and a big improvement to the module overall. I've made some in-line comments on the code, but generally it is of a very good standard.

I've run the 3 demo scripts we currently have on a limited set of runs that I have access to and they seem to produce mostly the same results as before. I expect any small differences are a result of bug fixes. Have you done regression testing on any other SOLPS cases you have?

There's a few places in the code with large commented-out blocks: these can be removed if they won't be added back in later. I'm thinking in particular in solps_plasma.py, where there are several commented-out plotting methods which should live in demo scripts instead.

Also, considering the amount of change here it would be worth adding to the changelog while the changes are still fresh. I'll defer to @CnlPepper on whether this is a big enough change to warrant a 1.2 (or even a 2.0) release, or whether a 1.1.1 would be suitable.

jacklovell commented:
One other comment: if Axisymmetric wrappers of  `SOLPSFunction2D` and `SOLPSVectorFunction2D` replace the corresponding 3D functions, perhaps we should remove the latter from the module entirely. The 3D functions are still present in this PR.

CnlPepper commented:
> One other comment: if Axisymmetric wrappers of `SOLPSFunction2D` and `SOLPSVectorFunction2D` replace the corresponding 3D functions, perhaps we should remove the latter from the module entirely. The 3D functions are still present in this PR.

I agree, it would be best to remove redundant code.

CnlPepper commented:
> Also, considering the amount of change here it would be worth adding to the changelog while the changes are still fresh. I'll defer to @CnlPepper on whether this is a big enough change to warrant a 1.2 (or even a 2.0) release, or whether a 1.1.1 would be suitable.

While there are large API changes to the SOLPSSimulation API, the core functionality for loading and generating Plasma objects 
 and the object layout are essentially the same. It is definitely a new minor release, possibly a major release.

The arguments for a major/minor version increment would be the degree of impact to the main usage of the package. This one lies on the border I suspect. As the active users of this package, you would be best placed to assess the impact on users and then decide if you want to go for v2.0.0 or v1.2.0.

Please do add a changelog. I would recommend getting into the habit of checking if it needs updating with every pull request.

Mateasek commented on cherab/solps/mesh_geometry.py:
```diff
@@ -40,89 +36,122 @@ class SOLPSMesh:
+                             must be 3 dimensional. Example shape is (4 x 32 x 98).
+                             In SOLPS notation: left/right - poloidal prev./next, bottom/top - radial prev./next.
+                             Cell indexing starts with 0 and -1 means no neighbour.
+    :param ndarray neighbix: Array of radial indeces of neighbouring cells in order: left, bottom, right, top,
```
shouldn't it be neighbiy?

Mateasek commented on cherab/solps/solps_plasma.py:
```diff
@@ -248,55 +597,201 @@ def eirene_simulation(self):
+
+        self._eirene = value
+
+    def b2_flux_to_velocity(self, poloidal_flux, radial_flux, parallel_velocity):
```
Wouldn't it be better to make this method a standalone instead of class method, like it is in the case of EFITEquilibrium class? This class is becoming quite big, it would help decreasing it. Also the method itself could be useful/used for various purposes. The same could be done with `eirene_flux_to_velocity`.

Mateasek commented:
That is a large amount of work @vsnever, very nice. Thanks for all the work you put into it, especially when SOLPS is not the easiest code to understand and read! I would like to propose one more change, since this is quite major rewrite. I would suggest also removing the commented plotting methods from the SOLPSimulation class. I think the plots would be better done outside in a separate file, as it is with equilibrium plots in Cherab. I don't mean that you should be also adding it now.

vsnever committed:
Fixes in response to reviewers comments. Removed redundant check for inside/outside mesh in SOLPSTotalRadiatedPower.

vsnever commented:
@jacklovell, @Mateasek, @CnlPepper, thank you very much for the code review. I hope I addressed all the issues you raised. For three of them I left the comments (see above). If I did not leave a comment, it means that the issue is fixed exactly as you suggested.

>I've run the 3 demo scripts we currently have on a limited set of runs that I have access to and they seem to produce mostly the same results as before. I expect any small differences are a result of bug fixes. Have you done regression testing on any other SOLPS cases you have?

When the data is read from the MDSPlus, the only difference should be in species velocities, because of the bug fix. When the data is read from the balance.nc file, then in addition to velocities, the density of impurity neutral atoms is different, because now it's obtained from the Eirene data. If the data is read from raw simulation files then the density of main plasma neutral atoms is also different because previously it was obtained from B2.

>Also, considering the amount of change here it would be worth adding to the changelog while the changes are still fresh. I'll defer to @CnlPepper on whether this is a big enough change to warrant a 1.2 (or even a 2.0) release, or whether a 1.1.1 would be suitable.

I updated the changelog suggesting a 1.2 version.

In addition to the improvements you suggested, I removed redundant check for inside/outside mesh in `SOLPSTotalRadiatedPower` emitter, and because now this emitter uses only a `total_radiation_f3d` from `SOLPSSimulation`, I think that passing the simulation object to this emitter is no longer needed, a Function3D instance is enough.

jacklovell reviewed (approved):
Looks good to me.

jacklovell commented:
@CnlPepper @Mateasek are you happy for these changes to be merged?

Mateasek commented:
Yes, sorry I am new to this, forgot to click on approve. Very nice piece of work!

CnlPepper commented:
Happy here, nice work!

vsnever commented:
@jacklovell, @Mateasek, @CnlPepper, thanks a lot!

@jacklovell, please do not close #34 when this is merged. There is still plenty of work to do.
